### PR TITLE
HashJoin should return Err when the right side input stream produce Err, add more join UTs to cover different join types

### LINKED ISSUE #4362: HashJoin should return Err when the right side input stream produce Err
**Describe the bug**

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by mingmwang**
# Which issue does this PR close?

Closes #4362.

# Rationale for this change

# What changes are included in this PR?

The PR also added some join UTs to cover different join types(Semi/Anti/Cross). It discloses several bugs and need to be addressed further. 

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by mingmwang** [comment]
@alamb @liukun4515

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 1]
```diff
@@ -1554,7 +1554,8 @@ impl HashJoinStream {
                     }
                     Some(result.map(|x| x.0))
                 }
-                other => {
+                Some(err) => Some(err),
```
Comment (liukun4515) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:1557]: 👍 this is also found in the refactor pr https://github.com/apache/arrow-datafusion/pull/4377/files#

File: datafusion/sql/src/planner.rs [hunk 2]
```diff
@@ -767,7 +767,8 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 let join_filter = filter.into_iter().reduce(Expr::and);
 
                 if left_keys.is_empty() {
-                    // When we don't have join keys, use cross join
+                    // TODO should not use cross join when the join_filter exists
+                    // https://github.com/apache/arrow-datafusion/issues/4363
```
Comment (jackwener) [inline datafusion/sql/src/planner.rs:771]: Perhaps our current crossjoin implementation is a little strange.
Some databases implement Crossjoin via Nest Loop Join. (cross join is logical, NLJ is physical)

**Review by alamb — APPROVED** [review]
Looks like a great improvement to me (the code looks good and the tests are 💯 ) Thank you @mingmwang

File: datafusion/sql/src/planner.rs (hunk 2 above)
Comment (alamb) [reply datafusion/sql/src/planner.rs:771]: 👍

**Pull request merged**